Tribes/src/app/_components/building-type/building-type.component.ts

```diff
@@ -91,13 +101,88 @@ export class BuildingTypeComponent implements OnInit {
 
   addNewBuilding() {
     if (this.goldAmount > COST_NEW_BUILDING) {
+      this.goldAmount = this.goldAmount - COST_NEW_BUILDING;
       this.buildingService.addNewBuilding(this.type);
     } else {
       this.errorMessage =  'You do not have enough money.';
     }
   }
 
   checkIfBuildingIsUpgradeable(level: number): boolean {
-    return this.getNumberOfBuildingsByLevel(level) !== 0 && this.townhallLevel > level && this.goldAmount >= this.getUpgradingCost(level);
+    return this.getNumberOfBuildingsByLevel(level) !== 0
+            && this.townhallLevel > level
+            && this.goldAmount >= this.getUpgradingCost(level)
+            && !this.checkIfAllBuildingByLevelIsUnderConstruction(level);
+  }
+
+  checkUpgradingConditions(level: number): string {
+    if (this.getNumberOfBuildingsByLevel(level) === 0) {
+      return 'You don\'t have any building of level ' + level;
+    } else if (this.checkIfAllBuildingByLevelIsUnderConstruction(level)) {
+      return 'All buildings of level ' + level + ' are upgrading';
+    } else {
+      if (this.townhallLevel <= level && this.goldAmount < this.getUpgradingCost(level)) {
+        return 'You don\'t have enough money and the townhall level is too low.';
+      } else if (this.townhallLevel <= level) {
+        return 'The townhall level is too low.';
+      } else if (this.goldAmount < this.getUpgradingCost(level)) {
+        return 'You don\'t have enough money';
+      }
+    }
+  }
+
+  upgradeBuilding(level: number): void {
+    if (this.type === 'townhall' && !this.checkIfTownhallIsUpgrading()) {
+      if (this.goldAmount >= this.getUpgradingCost(level)) {
```

Comment:
Here, that function could be reused i.e. `isUpgradePossible`

Reply:
By the townhall and by other buidlings the conditions to upgrade are not the same, therefore I think, that I can't use the same function.